Plane: added a new TAKEOFF flight mode

tridge opened the pull request:
takeoff then circle centered on the direction of takeoff. New parameters:

TKOFF_ALT: alt in meters above home for takeoff loiter
TKOFF_LVL_ALT: alt below which the wings are held level 
TKOFF_LVL_PITCH: min target pitch during initial takeoff (before TKOFF_LVL_ALT is reached)

All of the usual takeoff parameters are also active, including throttle delays, accel limits, taildragger controls etc

magicrub commented on ArduPlane/mode_takeoff.cpp:
```diff
@@ -0,0 +1,112 @@
+        SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, 100);
+        plane.takeoff_calc_roll();
+        plane.takeoff_calc_pitch();
+    } else {
```
would be cleaner if this was moved up with:
if (plane.flight_stage != AP_Vehicle::FixedWing::FLIGHT_TAKEOFF) {
        plane.calc_nav_roll();
        plane.calc_nav_pitch();
        plane.calc_throttle();
        return;
}
 and then the other takeoff stuff can happen below. As we add more features/logic to this, this part of the code will start to grow below this check

tridge replied:
I prefer it like it is

magicrub commented on ArduPlane/reverse_thrust.cpp:
```diff
@@ -71,6 +71,7 @@ bool Plane::allow_reverse_thrust(void) const
         allow |= (g.use_reverse_thrust & USE_REVERSE_THRUST_RTL);
         break;
     case Mode::Number::CIRCLE:
+    case Mode::Number::TAKEOFF:
```
why not use a new USE_REVERSE_THRUST_TAKEOFF?

tridge replied:
in what circumstances would you want reverse thrust for takeoff?

magicrub replied:
I can't think of one, which is why [it was never allowed in the first place](https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/reverse_thrust.cpp#L37-L40) so you're the one introducing it. I would prefer to see it just return false on it's own switch line.

tridge replied:
good point, fixed.
also fixed an issue with the reverse thrust bit handling code

magicrub commented on ArduPlane/mode_takeoff.cpp:
```diff
@@ -0,0 +1,112 @@
+    if (plane.flight_stage == AP_Vehicle::FixedWing::FLIGHT_TAKEOFF) {
+        SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, 100);
+        plane.takeoff_calc_roll();
+        plane.takeoff_calc_pitch();
```
takeoff_calc_pitch() has a check for MAV_CMD_NAV_TAKEOFF to prioritize roll over pitch. It needs to also check for this takeoff mode

tridge replied:
it does

magicrub commented on ArduPlane/mode_takeoff.cpp:
```diff
@@ -0,0 +1,112 @@
+        const float dist = target_dist;
+        const float alt = target_alt;
+        const float direction = degrees(plane.ahrs.yaw);
+
```
need to add crash detection stuff:
- crash_state.is_crashed = false;
- entry in in_preLaunch_flight_stage()

tridge replied:
thanks, fixed

tridge commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -638,7 +639,7 @@ void Plane::update_flight_stage(void)
                 set_flight_stage(AP_Vehicle::FixedWing::FLIGHT_NORMAL);
             }
-        } else {
+        } else if (control_mode != &mode_takeoff) {
```
having mode as a dumping ground isn't really any better than where it is

magicrub replied:
up until now, the flight stage only did anything in AUTO mode. So, I'm wondering if we should rethink it in general anyway..

tridge replied:
maybe, but not in this PR

magicrub commented on ArduPlane/mode_takeoff.cpp:
```diff
@@ -0,0 +1,114 @@
+    // @Increment: 1
+    // @Units: m
+    // @User: User
+    AP_GROUPINFO("DIST", 4, ModeTakeoff, target_dist, 200),
```
This needs to be lower. I'm thinking 100. Default loiter radius is 60.

JoshWelsh replied:
disagree, many flying fields don't like autonomous planes flying anti-pattern in the pattern.  I think it's best to let this go a little further out by default, and let the user bring it in if their condition allows.  Just my $0.02

magicrub replied:
by default, the plane will be orbiting with a min distance of 200+/-60 plus the 50 altitude. That's almost an effective 300m distance. That's a bit crazy. If I threw my plane up I would think it's flying away!!!!

JoshWelsh replied:
I flew this today.  Having the plane orbit with its furthest point away at almost 300m was not a concern, and also kept it out of the most immediate flying space where some others were toying around.

JoshWelsh commented:
I have a log and some feedback from the experience.. First and foremost, the plane took off nominally and was for the most part a non-event.  

Something I didn't particularly care for however was once the plane reached what I presume was TKOFF_LVL_ALT it then turned 45 or so degrees left.  I am suspecting this is the result of heading error accumulation as the wings are constrained to level, but I think a better experience here would be to reset the target shortly before the wings are allowed to roll over as the final climb to TKOFF_ALT begins.  

After allowing the plane to complete a couple loiters I then flew it southwest a bit, gained altitude, and simulated an accidental trigger again of takeoff mode, where the plane essentially "took off" again, climing a bit and then loitering in a new position.  I think this use case should have a different result, perhaps something like a check against isFlying() to decide whether or not to execute MAV_CMD_DO_TAKEOFF or perhaps just enter a loiter?

The log is at [this link](https://flynnsters-my.sharepoint.com/:u:/g/personal/josh_joshco_us/EQ0z3_YWurZDuCTViDOufloBii1jTGW21nosLCleRpSFpw?e=6ewOG0).

WickedShell commented:
>simulated an accidental trigger again of takeoff mode, where the plane essentially "took off" again, climbing a bit and then loitering in a new position. I think this use case should have a different result, perhaps something like a check against isFlying() to decide whether or not to execute MAV_CMD_DO_TAKEOFF or perhaps just enter a loiter?

I haven't reviewed the code, but this isn't the behavior I'd expect at all. If you are below the takeoff altitude then sure, re run the takeoff, but once above it we should always skip the takeoff and move into the loiter phase without having to do a climb.

vierfuffzig commented:
@tridge tested TKOFF mode on a testrig without mag. primary takeoff worked as expected. setting the loiter wp heading at motor start does not work without a mag though unless we have a GPS velocity vector at that time. setting the loiter wp heading once EKF yaw aligment is completed would add TKOFF support to no-mag setups.

tridge commented:
@vierfuffzig I've added support for TAKEOFF mode on boards with no compass. It re-calculates the loiter position once it reaches the target alt.
I've also added a check for the plane already flying, so TAKEOFF can be used like LOITER mode when flying

vierfuffzig commented:
awesome, thanks @tridge , will test asap!

tenten8401 commented:
I'll see if I can give it a go as well once I finish getting my wing to fly smoothly.

vierfuffzig commented:
@tridge works like s charm in SITL. hope weather allows for real life testing on sunday. thanks a lot!

tridge commented:
@vierfuffzig @tenten8401 great, thanks guys!

vierfuffzig commented:
@tridge might be nice to have TAKEOFF as an AUX FUNCTION too?

tridge commented:
@vierfuffzig yes, good idea

tridge commented:
merging this ready for plane 4.0
